### ISSUE TITLE
Building a better auth[nz]. A wishlist. Of sort.

### ISSUE THREAD

**Issue by AnrDaemon** [ForgeEssentials/ForgeEssentialsMain#1539]
First, a few key definitions.

1. Authentication: process of establishing the genuine identity of a subject.
2. Authorization: process of assigning (or matching against) permissions to the subject.

Note the difference. Many people confuse these two terms.
It is completely possible to authorize subject to perform some actions without establishing its identity.
Consider medieval-style granting of a royal seal ring. A person entrusted with such a ring is essentially authorized to act and speak on behalf of the monarch in situations appropriate to the seal grade entrusted.

With this subject cleared, let's move to the proposal.

Part I. Authentication.

Possible authentication options remain the same.
1. Anonimous identification. You just accept the connected user without checking its identity.
2. Mojang server auth token. (I'm not sure of the details, but I presume for simplicity that MJ provide actual authentication rather than only authorization.)
3. Local authentication. (Console /auth login <password>.)

The stuff in this part is fairly obvious. When the user logging in, it is considered anonimous, unless provided correct Mojang authentication credentials.
Server settings regarding authentication aren't complicated either. In fact, only one setting needs to be configured: password-auth ignore|optional|required.

Part II. Authorization.

When you are considering implementation of authorization system, the important part is to understand, that the level of trust to specific authentication schemes is up to the administrator's discretion, not implementor's.
As an example, an administrator may wish to allow all users a "spectator" mode, unless they are [previously] registered users, authenticated or not.
The scheme as I would like it to have is as follows:

The FE client is strictly optional. It enhance and streamline the process, but in no way it is a requirement.

The user states and possible values to use in authorization:
1. 'anonymous' - a user wasn't authenticated or otherwise identified. In most simple terms this means we are seeing this user first time.
2. 'registered' - we can identify that we have seen this user, but it did not authenticated explicitly.
  Possible values: set of {remote, local}. Remote, i.e. mojang auth. Local is a local registration facility.
3. 'authenticated' - the user was explicitly authenticated as him- or herself.
  Possible values: set of {remote, local}.

In regard to selected authorization group, the settings include:
1. Default group or action assigned to all users of the server. Must be set.
2. Default group assigned to all registered users. If not specified, fall back to previous option.
3. Default group assigned to all authenticated users. If not specified, fall back to previous option.

This set of options would allow for some hardened security schemes, like "allow all registered users build and stuff, but administrators must explicitly authenticate before using their powers to harm the server".
In situations of sharing the gaming system between multiple persons, this would prevent abuse of power, intentional or accidental.

Part III. Local registration.

While current system as it is implemented right now largely works as outlined above, when the FE client mod is installed, this allow for experience enhancement that is currently lacking.
A user authenticating on a server with FE client installed could receive a key from the server, that would provide user identification from now on.
The key is then stored on a server, associated with said user name. If the same user logging in later provide correct key, it is elevated to "registered" state automatically.
The local registration is not a substitution or a replacement for local authentication, but it may make users lives a little more convenient.

Part IV. Authorization grants.

Regardless of existence of local registration, the other desired part of the authorization is the ability to assign specific rights to the user implicitly.
This is commonly impelemented through "tokens".
Administrator create a token, assign specific actions to it, then pass the token to someone. That someone using the token and gain the associated rights.
Tokens may be volatile or persistent.
The only restriction I'd put on a token itself is its name spelling. It should be short enough to transfer verbally, and should not contain letters that can be misinterpreted. Hex string of something like sixteen symbols sounds reasonable.
Manual token naming is optional. I would say "discouraged", but it is up to server administrator's discretion.

**Comment by AnrDaemon** [comment]
I was intending to post it in response to another issue, but as I was writing it, the text was growing larger and larger. Then I decided to start again and rewrite it from scratch to be a little more consistent on its own.
And to the reply part… With no intention to push for a feature, I would like to voice my own opinion. And give some insight.
We have a community MC server running 1.7.10 with FE, populated by loosely tied society of forum members, their friends, their friends' friends, and given our history of running servers for different games, not only MC, for a couple of years, we've produced some rules.
The very first one is that the server is public. Anyone can login, register and play. But we will maintain authentication, and reserve rights to punish malicious activity.
Now, while we actively encourage people to use official client, the reality is such that the server that only allow official clients die out in a matter of weeks from sheer boredom. Playing on a server, where there's at most one person on-line at all times is no better than playing solo. Been there, tried that. With nearly all memory of a major i7 powerhouse taken up by an empty world no one care about is a titanic waste of electricity.
We're thankful for the community providing wonderful tools such as FE, allowing us to remain open to the people yet reasonable secure. However, the more straightforward authentication and/or authorization process is desirable. For both parts of the community.
I apologize for this little essay, but it was felt necessary to outline the environment.

**Comment by luacs1998** [comment]
TL;DR What do you want?

**Comment by AnrDaemon** [comment]
Everything! Plus a pony.
More seriously, there's 3 parts in the proposal that are currently unimplemented, and none of them dependent on any other.
Tokens. Client identification. Optional distinction between remote and local authentication.

**Comment by luacs1998** [comment]
Tokens: I'm sure I can speak to @olee if he can add a system to (temporarily) promote a user to a group, unti a time period expires. The admins can simply issue /p group addtemp <groupname> <playername> <duration>, and the player would be automatically removed from the group when the time expires.

As for "no rights until authenticated" that can be arranged too. Currently, the "authenticate" step is used for regular players, too, to prove their identity, but I think I'd prefer an /auth sudo kind of approach, which would require ops to key in their password AGAIN to prevent abuse.

Client identification: That's something like what I have in mind for #708. You authenticate once with a client, the client receives an (encrypted "garbage"!) token. Next time you connect, the client sends the token over to the server, the server decrypts it and identifies you. Of course you would need the FEClient installed, and you'd have to copy the token file between all the clients you want to use.

Remote/local authentication distinction: What I intend to do is to restructure auth-module into having auth "providers". For starters, you'd have the Mojang one (which happens automatically on player connection), and you'd have the FE one (/auth). I'm thinking of adding APIs to allow server admins to write their own ones (in java) that can link to an external website OpenID auth or something. But that's probably for another day.

**Comment by AnrDaemon** [comment]
Re: Temporary rights grant
I didn't hough of this option. I don't see a lot of uses for it, but if it can be an additional option, someone may find a use for it, I guess.
